File: Exercises/C/pi.c
/*

This program will numerically compute the integral of

                  4/(1+x*x) 
				  
from 0 to 1.  The value of this integral is pi -- which 
is great since it gives us an easy way to check the answer.

The is the original sequential program.  It uses the timer
from the OpenMP runtime library

History: Written by Tim Mattson, Nov 1999
         Start i from 0 instead of 1 and minor edits by Helen He, Nov 2020

*/

#include <stdio.h>
#include <omp.h>
static long num_steps = 1024*1024*1024;
double step;
int main()
{
   int i;
   double x, pi, sum = 0.0;
   double start_time, run_time;

   step = 1.0 / (double) num_steps;

   start_time = omp_get_wtime();

   for (i = 0; i < num_steps; i++) {
      x = (i + 0.5) * step;
      sum += 4.0 / (1.0 + x * x);
   }

   pi = step * sum;
   run_time = omp_get_wtime() - start_time;
   printf("pi = %lf, with %ld steps, in %lf secs\n ", pi, num_steps, run_time);
}

File: Exercises/Fortran/pi.f90
! This program will numerically compute the integral of
!                   4/(1+x*x) 
! from 0 to 1.  The value of this integral is pi -- which 
! is great since it gives us an easy way to check the answer.

! The is the original sequential program.  It uses the timer
! from the OpenMP runtime library

! History: C code written by Tim Mattson, 11/1999.
! adapted to Fortran code by Helen He, 09/2017. 

 
  PROGRAM MAIN
  USE OMP_LIB
  IMPLICIT NONE

  INTEGER :: i
  INTEGER, PARAMETER :: num_steps = 100000000
  REAL*8 :: x, pi, sum, step
  REAL*8 :: start_time, run_time

  sum = 0.0

  step = 1.0 / num_steps
  start_time = OMP_GET_WTIME()

  DO i = 1, num_steps
     x = (i - 0.5) * step
     sum = sum + 4.0 / (1.0 + x * x)
  ENDDO

  pi = step * sum
  run_time = OMP_GET_WTIME() - start_time

  WRITE(*,100) num_steps, pi,  run_time
100  FORMAT('pi with ',i14,' steps is ',f15.8,' in ',f8.3,' secs')

  END PROGRAM MAIN
